qutip/core/coefficient.py

```diff
@@ -273,30 +321,52 @@ def coeff_from_str(base, args, args_ctypes, compile_opt):
     hash_ = hashlib.sha256(bytes(parsed, encoding='utf8'))
     file_name = "qtcoeff_" + hash_.hexdigest()[:30]
     # See if it already exist, if not write and cythonize it
-    coeff = try_import(file_name, parsed)
+    coeff, file_name = try_import(file_name, parsed)
     if coeff is None or compile_opt['recompile']:
-        code = make_cy_code(parsed, variables, constants, raw, compile_opt)
+        code = make_cy_code(parsed, variables, constants,
+                            raw, compile_opt)
         coeff = compile_code(code, file_name, parsed, compile_opt)
     keys = [key for _, key, _ in variables]
     const = [fromstr(val) for _, val, _ in constants]
+    print(keys, variables)
     return coeff(base, keys, const, args)
 
 
 def try_import(file_name, parsed_in):
     """ Import the compiled coefficient if existing and check for
     name collision.
     """
+    get_root()
     coeff = None
     try:
         mod = importlib.import_module(file_name)
-        coeff = getattr(mod, "StrCoefficient")
-        parsed_saved = getattr(mod, "parsed_code")
-    except Exception as e:
-        parsed_saved = ""
-    if parsed_saved and parsed_in != parsed_saved:
-        # hash collision!
-        coeff = None
-    return coeff
+    except ModuleNotFoundError:
+        # Coefficient does not exist, to compile as file_name
+        return None, file_name
+
+    if mod.parsed_code == parsed_in:
+        # Coefficient found!
+        return mod.StrCoefficient, file_name
+
+    # hash collision!
+    # Increment filename
+    tries = -1
+    file_name_ = file_name
+    while mod.parsed_code != parsed_in and tries < 10:
```

Comment:
Good point, if the hash is not sufficient and we don't support infinite hash collision like `dict`, we should just quickly fail instead of trying to postpone. Expecting about 10000 coefficients used per users, 30 hex digit should be more than enough.